src/cls/IPM/Storage/Module.cls

```
@@ -85,6 +85,218 @@ Property AfterInstallMessage As %String(MAXLEN = "", XMLPROJECTION = "Element");
+        }
+        if primaryOnly {
+            set updateStep.HostName = ##class(%SYS.System).GetNodeName()
+            set updateStep.InstanceName = ##class(%SYS.System).GetInstanceName()
+        }
+        set updateStep.TimeStampEnd = $zdatetime($horolog,3)
+        $$$ThrowOnError(updateStep.%Save())
+    }
+    else {
+        if verbose {
+            write !, $char(9)_"Skipping seeding "_updateStep.MethodName_"() (already run or seeded)"
+        }
+    }
+}
+
+/// Executes an update step
+/// If an update step does not already exist, a new one gets created.
+/// newStepToApply - Gets marked as true to tell us to run all remaining steps in the list, whether or not they have been run/seeded before
+///                  Can carry this over to subsequent calls by passing the newStepToApply variable to those ExecuteUpdateStep() calls
+Method ExecuteUpdateStep(
+	updateStep As %IPM.Storage.UpdateStep.Base,
+	primaryOnly As %Boolean,
+	verbose As %Boolean = 0)
+{
+    // If it's a primary only step and this member is in the mirror but not a primary, don't execute it
+    if primaryOnly && $system.Mirror.IsMember() && '$system.Mirror.IsPrimary() {
+        if verbose {
+            write !, $char(9)_"Skipping "_updateStep.MethodName_"(). This step is to be run on the mirror primary only."
+        }
+        quit
+    }
+    if verbose {
+        write !, $char(9)_"Running "_updateStep.MethodName_"()"
+    }
+    if primaryOnly {
+        set updateStep.HostName = ##class(%SYS.System).GetNodeName()
+        set updateStep.InstanceName = ##class(%SYS.System).GetInstanceName()
+    }
+    set updateStep.TimeStampStart = $zdatetime($horolog,3)
+    set updateStep.Status = $$$OK
+    try {
+        $$$ThrowOnError($classmethod(updateStep.ClassName, updateStep.MethodName))
+        set updateStep.TimeStampEnd = $zdatetime($horolog,3)
+    } catch (ex) {
+        set updateStep.Status = ex.AsStatus()
```

If a single method errors, should throw the error to stop further processing